Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h

```diff
@@ -0,0 +1,131 @@
+  Care Ontario.
+
+==============================================================================*/
+
+#ifndef __qMRMLTextWidget_h
+#define __qMRMLTextWidget_h
+
+// SlicerQt includes
+#include "qSlicerWidget.h"
+
+// Text widgets includes
+#include "qSlicerTextsModuleWidgetsExport.h"
+#include "ui_qMRMLTextWidget.h"
+
+class vtkMRMLNode;
+class vtkMRMLTextNode;
+class qMRMLTextWidgetPrivate;
+
+/// \ingroup Slicer_QtModules_Texts
+class Q_SLICER_MODULE_TEXTS_WIDGETS_EXPORT qMRMLTextWidget : public qSlicerWidget
+{
+  Q_OBJECT
+
+public:
+  typedef qSlicerWidget Superclass;
+  qMRMLTextWidget(QWidget *parent=nullptr);
+  ~qMRMLTextWidget() override;
+
+  Q_PROPERTY(bool readOnly READ readOnly WRITE setReadOnly)
+  Q_PROPERTY(bool continuousUpdate READ continuousUpdate WRITE setContinuousUpdate)
+
+  /// Get the table node the columns are edited of.
+  Q_INVOKABLE vtkMRMLTextNode* mrmlTextNode() const;
+  Q_INVOKABLE vtkMRMLNode* mrmlNode() const;
+
+  /// Returns true if the text editor is read only
+  bool readOnly();
+
+  /// Returns true if the text editor is continuously updated from the text node and vice versa
+  bool continuousUpdate();
+
+public slots:
+
+  /// Reimplemented from qSlicerWidget
+  void setMRMLScene(vtkMRMLScene* scene) override;
+
+  /// Set the read only property of the text editor.
+  /// \sa readOnly()
+  void setReadOnly(bool readOnly);
+
+  /// Get the continuous update propery of the text editor
+  /// If true, the text editor will propogate the text to the vtkMRMLTextNode as it is modified, and vice versa.
+  /// If false, the text editor will only update the node when "Save" is clicked, and changes from the vtkMRMLTextNode will not be propogated
+  /// if the text is being edited.
+  /// \sa continuousUpdate()
+  void setContinuousUpdate(bool continuousUpdate);
+
+  /// Set the currently observed text node
+  /// \sa mrmlTextNode()
+  void setMRMLTextNode(vtkMRMLTextNode* textNode);
+
+  /// Utility function to simply connect signals/slots with Qt Designer
+  /// \sa mrmlNode()
+  void setMRMLNode(vtkMRMLNode* textNode);
+
+protected slots:
+  /// Update the GUI to reflect the currently selected text node.
+  void updateWidget();
+
+  /// Method invoked when the contents of the text node is modified
+  void onTextNodeModified();
+
+  /// Method invoked when the contents of the text edit is changed
+  void onTextChanged();
+
+  /// Method invoked when the "Edit" button is clicked
+  void onEditClicked();
+
+  /// Method invoked when the "Cancel" button is clicked
+  void onCancelClicked();
+
+  /// Method invoked when the "Save" button is clicked
+  void onSaveClicked();
+
+signals:
+  /// This signal is emitted if updates to the widget have finished.
+  void updateFinished();
+
+  /// This signal is emitted if the node changes
+  void mrmlNodeChanged(vtkMRMLNode*);
+
+  /// This signal is emitted if the read only propery is changed
+  void readOnlyChanged(bool);
```

Comment:
Instead of having a signal in the widget, should we instead have a a property in the text node ?

Reply:
I think that they would serve two different purposes, although having a readOnly property in the text node does seem like a useful feature.

- Read only in the qMRMLTextWidget allows text node content display without allowing editing (Whether or not the node is read only).

Reply:
Left ReadOnly out of the node for now.

Re a discussion with @lassoan, If it was implemented, it would probably be like the Locked parameter in markup and table nodes.